scripts/dynamic_prompts.py

```diff
@@ -0,0 +1,457 @@
+#!/usr/bin/env python
+
+"""
+Simple script to generate a file of InvokeAI prompts and settings
+that scan across steps and other parameters.
+"""
+
+import argparse
+import io
+import json
+import os
+import pydoc
+import re
+import shutil
+import sys
+import numpy as np
+from io import TextIOBase
+from itertools import product
+from pathlib import Path
+from multiprocessing import Process, Pipe
+from multiprocessing.connection import Connection
+from subprocess import PIPE, Popen
+from typing import Iterable, List
+
+import yaml
+from omegaconf import OmegaConf, dictconfig, listconfig
+
+
+def expand_prompts(
+    template_file: Path,
+    run_invoke: bool = False,
+    invoke_model: str = None,
+    invoke_outdir: Path = None,
+    processes_per_gpu: int = 1
+):
+    """
+    :param template_file: A YAML file containing templated prompts and args
+    :param run_invoke: A boolean which if True will pass expanded prompts to invokeai CLI
+    :param invoke_model: Name of the model to load when run_invoke is true; otherwise uses default
+    :param invoke_outdir: Directory for outputs when run_invoke is true; otherwise uses default
+    """
+    if template_file.name.endswith(".json"):
+        with open(template_file, "r") as file:
+            with io.StringIO(yaml.dump(json.load(file))) as fh:
+                conf = OmegaConf.load(fh)
+    else:
+        conf = OmegaConf.load(template_file)
+
+    # loading here to avoid long wait for help message
+    import torch
+    torch.multiprocessing.set_start_method('spawn')
+    gpu_count = torch.cuda.device_count() if torch.cuda.is_available() else 1
+    commands = expanded_invokeai_commands(conf, run_invoke)
+    children = list()
+    
+    try:
+        if run_invoke:
+            invokeai_args = [shutil.which("invokeai"),"--from_file","-"]
+            if invoke_model:
+                invokeai_args.extend(("--model", invoke_model))
+            if invoke_outdir:
+                invokeai_args.extend(("--outdir", os.path.expanduser(invoke_outdir)))
+
+            processes_to_launch = gpu_count * processes_per_gpu
+            print(f'>> Spawning {processes_to_launch} invokeai processes across {gpu_count} CUDA gpus', file=sys.stderr)
+            import ldm.invoke.CLI
+            parent_conn, child_conn = Pipe()
+            children = set()
+            for i in range(processes_to_launch):
+                p = Process(target=_run_invoke,
+                            args=(child_conn,
+                                  parent_conn,
+                                  invokeai_args,
+                                  i%gpu_count,
+                                  )
+                            )
+                p.start()
+                children.add(p)
+            child_conn.close()
+            sequence = 0
+            for command in commands:
+                sequence += 1
+                parent_conn.send(command+f' --fnformat=dp.{sequence:04}.{{prompt}}.png')
```

Comment:
because filenames include the prompt, `[negative prompts]` and `<embeddings>` break file saving. I hacked around it here by replacing `{prompt}` with a random base32 value, but didn't find a way to control it better.

Reply:
The `--fnformat` argument is really old code, but it is supposed to be escaping bad characters. I'll investigate.

Reply:
This is now fixed. The filenames are written out to disk with the <>[] characters intact and do not cause a crash.